File: NS_multicomp.m

% % Copyright (c) 2019-2020 Carlo Manzo & Tina Košuta 

% If you use this code, please cite: 
% (http://dx.doi.org/10.1039/C9CP05616E)
% (https://arxiv.org/abs/1909.13133)

% Permission is hereby granted, free of charge, to any person 
% obtaining a copy of this software and associated documentation files 
% (the "Software"), to deal in the Software without restriction, 
% including without limitation the rights to use, copy, modify, merge, 
% publish, distribute, sublicense, and/or sell copies of the Software, 
% and to permit persons to whom the Software is furnished to do so, 
% subject to the following conditions:

% The above copyright notice and this permission notice shall be 
% included in all copies or substantial portions of the Software.

% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
% EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
% MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
% IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
% CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
% TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
% SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

%%

clear all
close all
clc

%% set param for Nested Sampling
warning off
% Number of particles
NSpar.N =  30;% 20
% Depth to go to
NSpar.depth =  40;

NSpar.pdfmodel='lognormal'; 
switch NSpar.pdfmodel
    case 'lognormal'
        NSpar.pdf=@(x,mu,sigma) 0.5*(erf( (mu - log(x-1))./(sqrt(2)*sigma) ) -erf( (mu- log(x))./(sqrt(2)*sigma) ) );
        NSpar.pdfpar = [3.3491 0.8462];    
    %    add extra cases
end

NSpar.delt=1.5;

%% load data

locs=readtable('sample_data.txt');
locs=table2array(locs(:,2));
locs=locs(isnan(locs)==0);

%
%% create x
locx=unique(locs);
ll=1:1:3*max(locx); % to make sure of normalization after numerical convolution
n=histc(locs,locx)'; % histogram of localizations

% create calibration function fo monomers
if length(NSpar.pdfpar)==2
    y(:,1)=NSpar.pdf(ll,NSpar.pdfpar(1),NSpar.pdfpar(2));
    % to use one parameter distributions
%elseif length(NSpar.pdfpar)==1 
%    y(:,1)=NSpar.pdf(ll,NSpar.pdfpar(1));
end

% condit to stop
flagmax=0;
% number of species
num_params=1;

% start adding one extra component at the time
while    flagmax<1   % condition on the maximum of logZnew
    num_params = num_params+1;
    
    %% generate random particles (sum = 1)
    particles=dirichletrnd(NSpar.delt*ones(1,num_params),NSpar.N);
    %% add a convolution term
    y(:,num_params)=ifft(fft(y(:,1)).^[num_params]);
    y1=y(locx,:);
    %%
    % calculate log_prior
    logps=logdirichlet(particles, NSpar.delt); % dirichlet
    % sum function (sum(alpha*f)) for each particle
    fmix=particles*y1';  % or mix=y*particles';
    
    % calculate log_likelihood on every particle
    %logls= sum((n).*log(fmix),2);
    logls=(n*log(fmix'))';
    % initialize NS parameters and evidence
    logwidth=0;
    logZnew=0;
    %%
    keep=[];
    keep2=[];
    % initialize ratio of evidence for convergence
    Zrat=Inf;
    i=0;
    while Zrat>log(1E-5)
        i=i+1;
        % Find worst particle
        [worst_logl ,worst] = min(logls);
        %calculate logweight times loglikelihood
        logwidth=-i*log(1+1/NSpar.N) -log(NSpar.N) +worst_logl  ;
        
        % dead particles
        keep = [keep; particles(worst,:) logls(worst)];
        keep2=[keep2; logwidth];
        
        % set loglikelihood threshold
        threshold = logls(worst);
        
        % Copy a surviving particles
        flag=0;
        while flag==0
            which = randsample(NSpar.N, 1);
            if(which ~= worst)
                break
            end
        end
        particles(worst,:) = particles(which,:);
        % upgrade likelihood and prior 
        logps(worst) = logps(which);
        logls(worst) = logls(which);
        
        % evolve particle with MCMC
        sig=.1;% tune step width based on rejection 
        accepted=0; % initialize accepted
        R=0;
        for j=1:NSpar.depth
            new0=particles(worst,:);
 
             kk=randsample(num_params, 1);
             step_size=random('norm',0,sig,1);
%             % periodic boundaries in [0,1]             
             new0(1,kk)=new0(1,kk) + step_size;
             new0(1,kk)=new0(1,kk)-floor(new0(1,kk));       
             new0(1,:)=new0(1,:)./sum(new0(1,:));

                 
                fmix_new=new0*y1';  % or mix=y*particles';
                logl_new=(n*log(fmix_new'))';% log likelihood
                logp_new=logdirichlet(new0(1,:), NSpar.delt);% prior
                
                % Metropolis-Hastings
                if logl_new > threshold && rand<=exp(min([logp_new-logps(worst),0]))
                    particles(worst,:) = new0(1,:);
                    logls(worst) = logl_new;
                    logps(worst) = logp_new;
                    accepted = accepted + 1;
                else
                  R=R+1;  
                end
                
                if accepted>R
                    sig=sig*exp(1/accepted);
                elseif accepted<R
                    sig=sig/exp(1/R);
                end
              
            %end
        end %end MCMC
        clear fmix_new logl_new logp_new 
        
        logZnew=(log(sum( exp( keep2-max(keep2) )) ) + max(keep2));
        logZres=-i*log(1+1/NSpar.N) -log(NSpar.N) +  (log(sum( exp( logls-max(logls) )) ) + max(logls)); 
        
        
        Zrat=(logZres   -  logZnew);
        
    end
    
    %%
    clear Zrat fmix logZnew logZres logwidth new0 z z0 worst worst_logl 
    
    logwidth2=-i*log(1+1/NSpar.N) -log(NSpar.N)  + logls; %  
    
    % add dead particles
    keep2=[keep2; logwidth2];
    keep=[keep; particles logls];
    
    clear particles logls logwidth2 logps sig 
    
    % calculate Evidence
    NSout.logZnew(num_params-1)=(log(sum( exp( keep2-max(keep2) )) ) + max(keep2));
    
   % posterior weights (normalized) 
    post_weights = exp( keep2   - NSout.logZnew(num_params-1));
    
    
    % entropy
    NSout.ent(num_params-1) = -sum(post_weights.*log(post_weights + 1E-300));
    
    % effective posterior sample size
    NSout.ess(num_params-1) = floor(exp(NSout.ent(num_params-1)));
    
    % Information
    NSout.H(num_params-1) =  sum(post_weights.*(keep(:, size(keep,2)) -  NSout.logZnew(num_params-1)));
    
    % Error of logEvidence
    NSout.err(num_params-1) = sqrt(NSout.H(num_params-1)./NSpar.N);
    if NSout.H(num_params-1)<0
        return
    end
    
    
    %% weighted mean
    NSout.alphas{num_params-1}=sum(keep(:,1:end-1).*post_weights);
    NSout.erralpha{num_params-1}= sqrt(sum( (keep(:,1:end-1) -NSout.alphas{num_params-1}).^2 .*post_weights));
            
    % check for max conditions to stop model
    [m1,in1]=max(NSout.logZnew);

    if num_params-1>2
        if in1<num_params-1
            flagmax=1;
        end
    end
    clear m1 in1
    
end
%% display parameters corresponding to max logZnew
[NSout.logZ,imaxZ]=max(NSout.logZnew);
%
NSout_r.NbestZ=imaxZ+1;
% 
NSout_r.alphas=NSout.alphas{imaxZ};
%
NSout_r.erralpha=NSout.erralpha{imaxZ};
%
NSout_r.err=NSout.err(imaxZ);
%
NSout_r.ent=NSout.ent(imaxZ);
NSout_r.ess=NSout.ess(imaxZ);
NSout_r.H=NSout.H(imaxZ);
NSout_r
 

File: NS_multicomp.R
# # Copyright (c) 2019-2020 Carlo Manzo & Tina Košuta 
#
# If you use this code, please cite: 
# (http://dx.doi.org/10.1039/C9CP05616E)
# (https://arxiv.org/abs/1909.13133)
#
# Permission is hereby granted, free of charge, to any person 
# obtaining a copy of this software and associated documentation files 
# (the "Software"), to deal in the Software without restriction, 
# including without limitation the rights to use, copy, modify, merge, 
# publish, distribute, sublicense, and/or sell copies of the Software, 
# and to permit persons to whom the Software is furnished to do so, 
# subject to the following conditions:
  
# The above copyright notice and this permission notice shall be 
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#Upload data
data <- read.table("sample_data.txt")
data <- as.numeric(data[,]) #optional; number of localizations is needed as vector
nb <- table(data)

#mean of logN
p1_1 = 3.3491 

#sd of logN
p1_2 = 0.8462  

############### FUNCTIONS

library(gtools) #rdirichlet
library(pracma) #for erf and fmincom

#calibration curve
pp <- function(x,mu,sigma)  { 0.5*(erf((mu-log(x-1))/(sqrt(2)*sigma))-erf((mu-log(x))/(sqrt(2)*sigma))) }

#log likelihood
log_likelihood <- function(data, theta){
  lambda=ft_u_mat%*%theta
  logL=data%*%log(lambda) 
  return(logL)
}

############### MODEL PARAMETERS

# Number of particles
N <- 30

# MCMC steps per NS iteration
mcmc_steps <- 40

# delta - shape of Dirichlet distribution
delta <- 1.5

# maximum number of parameters in the model we expect
num_par <- 8

############### NESTED SAMPLING ALGORITHM

results <- list()

for(Nmax in 2:num_par){
  
    #calibration
    ft_u_mat <- NULL
    pp_ll <- seq(1, 3*max(data) ,1)
    pp_ft_mat <- matrix(0, ncol=Nmax, nrow=pp_ll)
    pp_ft <- pp(pp_ll, p1_1, p1_2)
    pp_ft_mat <- sapply(1:Nmax, function(x) Re(fft((fft(pp_ft)^x)/length(pp_ft), inverse=TRUE)))
    
    ll_u <- sort(unique(data)) # subset of our values (for our data)
    ft_u_mat <- pp_ft_mat[ll_u,]
    
    #deltas for each parameter in Dirichlet
    delta_dir <- rep(delta, Nmax) #for uniform prior
    
    points <- rdirichlet(N, delta_dir) #each column is new particle
    logL <- sapply(1:N, function(x) log_likelihood(data=nb, theta=points[x,]))
    logP <- log(ddirichlet(points, delta_dir))
    
    logwt <- 0 #Outermost interval of prior mass
    logZnew <- 0
    keep <- list()
    keep2 <- NULL
    Zrat <- Inf
    j <- 0
    
    while(Zrat > log(1e-5)){    #turn the inequallity to create condition
      
      j <- j+1
      
      min <- which.min(logL) #lowest likelihood
      worst_logl <- logL[min]
      logwt <- (-j*log(1+1/N) -log(N)) +worst_logl #Calculate weight of worst walker
      
      #data of first walker, Lmin, logZ, w
      keep[[j]] <- c(points[min,], logL[min])
      keep2[j] <- logwt
      
      threshold <- logL[min]
      
      Wnew <- sample(as.numeric(1:N)[-min],1) #remove walker with logL min and
      points[min,] <- points[Wnew,] #copy new walker; then walk the coppied walker on the place of walker min
      logL[min] <- logL[Wnew]
      logP[min] <- logP[Wnew]
      
      sig <- 0.1
      accept <- 0
      
      for (i in 1:mcmc_steps){
        
        k <- sample(1:Nmax,1)
        
        new <- points[min,]
        step_size <- min(rnorm(1,0,sig), 1-1e-6)
        new[k] <- ifelse(new[k]+step_size>=1, 1-(new[k]+step_size-1), abs(new[k]+step_size))
        new <- new/sum(new)
  
        lambda_new <- ft_u_mat%*%new
        logl_new <- nb%*%log(lambda_new) #the only changing part of likelihood; LOG LIIKELIHOOD
        logP_new <- log(ddirichlet(matrix(new, nrow=1), delta_dir))
        logPdif <- logP_new-logP[min]
        
        if(logl_new >= threshold && runif(1) <= exp(logPdif)){
          points[min,] <- new
          logL[min] <- logl_new
          logP[min] <- logP_new
          accept  <-  accept+1
        }
        
        sig <- ifelse(accept>(i-accept), sig*exp(1/accept), sig/exp(1/(i-accept)))
        
        }
      
      logZnew <- log(sum(exp(keep2-max(keep2)))) + max(keep2)  #new logZ
      logZrem <- -j*log(1+1/N)-log(N)+(log(sum(exp(logL-max(logL)))) + max(logL))  #remaining evidence
      Zrat <- (logZrem-logZnew)
      
    }
    
      logws <- -(1:j)*log(1+1/N) #posterior weights
      logwt2 <- (-j*log(1+1/N)) + logL #calculate the weight of each walker with new voulmes
      
      keep2 <- c(keep2, logwt2) 
      
      keep <- do.call(rbind, keep)
      keep <- rbind(keep, cbind(points,logL))  #add remaining walkers
      
      # Calculate marginal likelihood
      logZ <- log(sum(exp(keep2-max(keep2)))) + max(keep2) #final logZ
      
      post_weights <- exp(keep2 - logZ) #normalised posterior weights
      
      #H calculation
      H <- sum(post_weights*(keep[,ncol(keep)]-logZ))
  
      #Error
      error <- sqrt(H/N)
      
      #N samples from posterior
      ent <- -sum(post_weights*log(post_weights + 1E-300))
      Npost <- floor(exp(ent)) #effective size for posterior samples
      
      # Create posterior samples by resampling; importance weigth
      posterior_samples_ <- list()
      posterior_weights <- NULL
      
      # Counter
      k <- 1
      top <- max(post_weights)
      while(k<=Npost){
        # Choose one of the samples
        which <- sample(1:nrow(keep), 1)
        # Acceptance probability
        prob <- post_weights[which]/top
        if(runif(1) <= prob){
          posterior_samples_[[k]] <- keep[which, ]
          posterior_weights[k] <- post_weights[which]
          k <- k + 1
        }
      }
      
      posterior_samples <- do.call(rbind, posterior_samples_)
      
      #weighted mean from posterior sample
      Wm <- apply(posterior_samples[,-ncol(posterior_samples)]*posterior_weights,2,sum)/sum(posterior_weights)

    results[[Nmax]] <- list("N"=Nmax,"logZ"=logZ,"Wm"=Wm, "ent"= ent, "Npost"=Npost, "H"=H, "error"=error)      
}

#compare logZ for all models
logZ_results <- data.frame("Number of parameters"=sapply(2:num_par, function(x) results[[x]][["N"]]),
                            "logZ"=sapply(2:num_par, function(x) results[[x]][["logZ"]]))


#print results for model with largest logZ
results[[which.max(logZ_results[,"logZ"])]]

